src/modules/bulletproofs/rangeproof_uncompressed_impl.h

```diff
@@ -0,0 +1,146 @@
+
+/* Prover context data:
+ *    bytes 0-32:   x (hash challenge)
+ *    bytes 32-64:  y (hash challenge)
+ *    bytes 64-96:  z (hash challenge)
+ *    bytes 96-160: lr_generator data
+ */
+
+/* Step 0 of the proof.
+ * Uses the value but not the blinding factor. Takes the complete commitment,
+ * but only to put it into the hash.
+ * Outputs the points A and S, encoded in 65 bytes (one byte with A's parity
+ * in the LSB, S's parity in the second-LSB, 32-byte A.x, 32-byte S.x).
+ * Updates the state to include the hashes y and z.
+ */
+static int secp256k1_bulletproofs_rangeproof_uncompressed_prove_step0_impl(
+    const secp256k1_ecmult_gen_context* ecmult_gen_ctx,
+    secp256k1_bulletproofs_prover_context* prover_ctx,
+    unsigned char* output,
+    const size_t n_bits,
+    const uint64_t value,
+    const uint64_t min_value,
+    const secp256k1_ge* commitp,
+    const secp256k1_ge* asset_genp,
+    const secp256k1_bulletproofs_generators* gens,
+    const unsigned char* nonce,
+    const secp256k1_scalar* enc_data,
+    const unsigned char* extra_commit,
+    size_t extra_commit_len
+) {
+    secp256k1_sha256 sha256;
+    unsigned char commit[32];
+    secp256k1_scalar alpha, rho;
+    secp256k1_scalar tmp_l, tmp_r;
+    secp256k1_gej gej;
+    secp256k1_ge ge;
+    size_t i;
+    int overflow;
+
+    memset(prover_ctx->data, 0, sizeof(prover_ctx->data));
+    memset(output, 0, 65);
+
+    /* Sanity checks */
+    if (n_bits > 64) {
+        return 0;
+    }
+    if (gens->n < n_bits * 2) {
+        return 0;
+    }
+    if (value < min_value) {
+        return 0;
+    }
+    if (n_bits < 64 && (value - min_value) >= (1ull << n_bits)) {
+        return 0;
+    }
+    if (extra_commit_len > 0 && extra_commit == NULL) {
+        return 0;
+    }
+
+    /* Commit to all input data: min value, pedersen commit, asset generator, extra_commit
+     * Pass the output in as scratch space since we haven't used it yet. */
+    secp256k1_bulletproofs_commit_initial_data(commit, output, n_bits, min_value, commitp, asset_genp, extra_commit, extra_commit_len);
+
+    /* Compute alpha and rho, adding encrypted data to alpha (effectively adding
+     * it to mu, which is one of the scalars in the final proof) */
+    secp256k1_scalar_chacha20(&alpha, &rho, nonce, 0);
+    secp256k1_scalar_add(&alpha, &alpha, enc_data);
+
+    /* Compute and output A */
+    secp256k1_ecmult_gen(ecmult_gen_ctx, &gej, &alpha);
+    for (i = 0; i < n_bits; i++) {
+        secp256k1_ge aterm = gens->gens[2 * i + 1];
+        size_t al = !!((value - min_value) & (1ull << i));
+
+        secp256k1_ge_neg(&aterm, &aterm);
+        secp256k1_fe_cmov(&aterm.x, &gens->gens[2 * i].x, al);
+        secp256k1_fe_cmov(&aterm.y, &gens->gens[2 * i].y, al);
+        secp256k1_gej_add_ge(&gej, &gej, &aterm);
+    }
+    secp256k1_ge_set_gej(&ge, &gej);
+    secp256k1_fe_normalize_var(&ge.x);
+    secp256k1_fe_normalize_var(&ge.y);
+    output[0] = secp256k1_fe_is_odd(&ge.y) << 1;
```

Comment:
I don't think this would save us anything. We would still need to call `normalize_var` so many times; we'd have to add another `secp256k1_ge` variable which would increase stack space and negate the lines-of-code savings; and it would require the user to mentally keep track of `A` while `S` is being computed.